### DISCUSSION TITLE
Using eUPF with Raspberry Pi 5

### DISCUSSION THREAD

**Discussion by howardswe499** [edgecomllc/eupf#536] [Q&A] [answered]
Hi All,

I am trying to build and run eUPF from sources on a Raspberry Pi 5, running Raspbian Bookworm Kernel 6.1

I have confirmed that the kernel's CONFIG_BPF & CONFIG_BPF_SYSCALL are set to "y".

I am running into this error message in compilation step:

```
user@howard-pi:~/eupf# BPF_CFLAGS="-DENABLE_LOG" go generate -v ./cmd/...
...
cmd/ebpf/objects.go
In file included from /home/user/eupf/cmd/ebpf/xdp/n3n6_entrypoint.c:17:
In file included from /usr/include/linux/bpf.h:11:
/usr/include/linux/types.h:5:10: fatal error: 'asm/types.h' file not found
#include <asm/types.h>
         ^~~~~~~~~~~~~
1 error generated.
Error: can't execute clang: exit status 1
exit status 1
cmd/ebpf/objects.go:23: running "go": exit status 1
...
```
One thing to note is that I installed `gcc-multilib-x86-64-linux-gnu` as a substitute for `gcc-multilib`, as there was no arm64 version in the default repositories.

I'd like to know:
1. If there are other substitutes for gcc-multilib that would work for eUPF and if I could be pointed to it
2. Is there a workaround instead

Many Thanks
Howard

**Comment by pirog-spb** [comment, marked as answer]
Hi Howard @howardswe499,

I've successfully built eUPF on MacOS (arm) and there was the same issue: no gcc-multilib for arm arch.

What I've made to overcome the issue:

1. install gcc package instead of gcc-multilib
2. install linux headers manually
3. update symlinks to make it works

```
sudo apt install --no-install-recommends -y clang llvm gcc libbpf-dev
sudo apt install linux-headers-arm64
sudo ln -s /usr/include/aarch64-linux-gnu/asm/ /usr/include/asm
sudo ln -s /usr/include/aarch64-linux-gnu/sys/ /usr/include/sys
sudo ln -s /usr/include/aarch64-linux-gnu/bits/ /usr/include/bits
sudo ln -s /usr/include/aarch64-linux-gnu/gnu/ /usr/include/gnu
```

Hope it may help.

-- BR, Alex

**Reply by howardswe499** [reply]
Hi Alex @pirog-spb ,

Your solution worked perfectly. Thanks a lot.

BR
Howard

**Reply by krezefal** [reply]
> sudo apt install --no-install-recommends -y clang llvm gcc libbpf-dev
> sudo apt install linux-headers-arm64
> sudo ln -s /usr/include/aarch64-linux-gnu/asm/ /usr/include/asm
> sudo ln -s /usr/include/aarch64-linux-gnu/sys/ /usr/include/sys
> sudo ln -s /usr/include/aarch64-linux-gnu/bits/ /usr/include/bits
> sudo ln -s /usr/include/aarch64-linux-gnu/gnu/ /usr/include/gnu

In my case (Ubuntu 24.04) there was an error on the 2nd command, especially `sudo apt install linux-headers-arm64`:

```
E: Package 'linux-headers-arm64' has no installation candidate
```

Instead, I have successfully used this one:

```
sudo apt install linux-headers-oracle-64k
```
With the further symlinks creation according to comment above from @pirog-spb
